staging/src/k8s.io/apiserver/pkg/apis/apiserverinternal/validation/validation.go

```diff
@@ -0,0 +1,122 @@
+	return allErrs
+}
+
+func validateServerStorageVersion(ssv apiserverinternal.ServerStorageVersion, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	for _, msg := range apimachineryvalidation.NameIsDNSSubdomain(ssv.APIServerID, false) {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("apiServerID"), ssv.APIServerID, msg))
+	}
+	if errs := utilvalidation.IsDNS1035Label(ssv.EncodingVersion); len(errs) > 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("encodingVersion"), ssv.EncodingVersion, strings.Join(errs, ",")))
+	}
+
+	found := false
+	for i, dv := range ssv.DecodableVersions {
+		if errs := utilvalidation.IsDNS1035Label(dv); len(errs) > 0 {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("decodableVersions").Index(i), dv, strings.Join(errs, ",")))
+		}
+		if dv == ssv.EncodingVersion {
+			found = true
+		}
+	}
+	if !found {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("decodableVersions"), ssv.DecodableVersions, fmt.Sprintf("decodableVersions must include encodingVersion %s", ssv.EncodingVersion)))
+	}
+	return allErrs
+}
+
+func commonVersion(ssv []apiserverinternal.ServerStorageVersion) *string {
+	if len(ssv) == 0 {
+		return nil
+	}
+	commonVersion := ssv[0].EncodingVersion
+	for _, v := range ssv[1:] {
+		if v.EncodingVersion != commonVersion {
+			return nil
+		}
+	}
+	return &commonVersion
+}
+
+func validateCommonVersion(svs apiserverinternal.StorageVersionStatus, fldPath *field.Path) *field.Error {
+	actualCommonVersion := commonVersion(svs.StorageVersions)
+	if actualCommonVersion == nil && svs.CommonEncodingVersion == nil {
+		return nil
+	}
+	if actualCommonVersion == nil && svs.CommonEncodingVersion != nil {
+		return field.Invalid(fldPath.Child("commonEncodingVersion"), *svs.CommonEncodingVersion, "should be nil if servers do not agree on the same encoding version, or if there is no server reporting the supported versions yet")
+	}
+	if actualCommonVersion != nil && svs.CommonEncodingVersion == nil {
+		return field.Invalid(fldPath.Child("commonEncodingVersion"), svs.CommonEncodingVersion, fmt.Sprintf("the common encoding version is %s", *actualCommonVersion))
+	}
+	if actualCommonVersion != nil && svs.CommonEncodingVersion != nil && *actualCommonVersion != *svs.CommonEncodingVersion {
+		return field.Invalid(fldPath.Child("commonEncodingVersion"), *svs.CommonEncodingVersion, fmt.Sprintf("the actual common encoding version is %s", *actualCommonVersion))
+	}
+	return nil
+}
+
+func validateStorageVersionCondition(conditions []apiserverinternal.StorageVersionCondition, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	// We do not verify that the condition type or the condition status is
+	// a predefined one because we might add more type or status later.
+	for i, condition := range conditions {
+		for _, msg := range validation.IsQualifiedName(string(condition.Type)) {
```

Comment:
need to validate that type is unique among all conditions, since it is the merge key